library/ssl_tls13_server.c

```diff
@@ -1762,9 +1763,42 @@ static void ssl_tls13_update_early_data_status(mbedtls_ssl_context *ssl)
         return;
     }
 
-    /* We do not accept early data for the time being */
     ssl->early_data_status = MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED;
 
+    if (ssl->conf->early_data_enabled == MBEDTLS_SSL_EARLY_DATA_DISABLED) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            1, ("EarlyData: rejected. configured disabled."));
+        return;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG(
+        3, ("EarlyData: conf->max_early_data_size = %u",
+            (unsigned int) ssl->conf->max_early_data_size));
+
+    if (!mbedtls_ssl_conf_tls13_some_psk_enabled(ssl)) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            1,
+            ("EarlyData: rejected. psk or psk_ephemeral is not available."));
+        return;
+    }
+
+    if (handshake && handshake->resume != 1) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            1, ("EarlyData: rejected. not resumption session."));
+        return;
+    }
+
```

Comment:
Here, I think we should quote the spec (section 4.2.10):
```
In order to accept early data, the server MUST have accepted a PSK
cipher suite and selected the first key offered in the client’s
"pre_shared_key" extension. In addition, it MUST verify that the
following values are the same as those associated with the
selected PSK:
-The TLS version number
-The selected cipher suite
-The selected ALPN [RFC7301] protocol, if any
```
then, check the version number, that it is the first PSK that was selected (`ssl->handshake->selected_identity` can be used for that I think) as well as the ciphersuite. Then have a comment that ALPN is not checked yet.

Otherwise, related to that, I think in `ssl_tls13_offered_psks_check_identity_match_ticket()` we should reject straight away TLS 1.2 tickets. Currently they are rejected by the `ticket_flags` check but I'd prefer to make that more explicit.

Reply:
I just move client side check to EE message. ServerHello message won't send early data extension.

Reply:
> I just move client side check to EE message. ServerHello message won't send early data extension.

Yes, good point.